pkg/controller/cluster/cluster_controller_phases.go

```diff
@@ -0,0 +1,212 @@
+			return nil, nil
+		} else if apierrors.IsNotFound(err) {
+			return nil, errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second},
+				"could not find %v %q for Cluster %q in namespace %q, requeuing",
+				ref.GroupVersionKind(), ref.Name, cluster.Name, cluster.Namespace)
+		}
+		return nil, err
+	}
+
+	objPatch := client.MergeFrom(obj.DeepCopy())
+
+	// Delete the external object if the Cluster is being deleted.
+	if !cluster.DeletionTimestamp.IsZero() {
+		if err := r.Delete(ctx, obj); err != nil {
+			return nil, errors.Wrapf(err,
+				"failed to delete %v %q for Cluster %q in namespace %q",
+				obj.GroupVersionKind(), ref.Name, cluster.Name, cluster.Namespace)
+		}
+		return obj, nil
+	}
+
+	// Set external object OwnerReference to the Cluster.
+	ownerRef := metav1.OwnerReference{
+		APIVersion: cluster.APIVersion,
+		Kind:       cluster.Kind,
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+
+	if !util.HasOwnerRef(obj.GetOwnerReferences(), ownerRef) {
+		obj.SetOwnerReferences(util.EnsureOwnerRef(obj.GetOwnerReferences(), ownerRef))
+		if err := r.Patch(ctx, obj, objPatch); err != nil {
+			return nil, errors.Wrapf(err,
+				"failed to set OwnerReference on %v %q for Cluster %q in namespace %q",
+				obj.GroupVersionKind(), ref.Name, cluster.Name, cluster.Namespace)
+		}
+	}
+
+	// Add watcher for external object, if there isn't one already.
+	_, loaded := r.externalWatchers.LoadOrStore(obj.GroupVersionKind().String(), struct{}{})
+	if !loaded && r.controller != nil {
+		klog.Infof("Adding watcher on external object %q", obj.GroupVersionKind())
+		err := r.controller.Watch(
+			&source.Kind{Type: obj},
+			&handler.EnqueueRequestForOwner{OwnerType: &clusterv1.Cluster{}},
+		)
+		if err != nil {
+			r.externalWatchers.Delete(obj.GroupVersionKind().String())
+			return nil, errors.Wrapf(err, "failed to add watcher on external object %q", obj.GroupVersionKind())
+		}
+	}
+
+	// Set error reason and message, if any.
+	errorReason, errorMessage, err := external.ErrorsFrom(obj)
+	if err != nil {
+		return nil, err
+	}
+	if errorReason != "" {
+		clusterStatusError := common.ClusterStatusError(errorReason)
+		cluster.Status.ErrorReason = &clusterStatusError
+	}
+	if errorMessage != "" {
+		cluster.Status.ErrorMessage = pointer.StringPtr(errorMessage)
+	}
+
+	return obj, nil
+}
+
+// reconcileInfrastructure reconciles the Spec.InfrastructureRef object on a Cluster.
+func (r *ReconcileCluster) reconcileInfrastructure(ctx context.Context, cluster *v1alpha2.Cluster) error {
+	if cluster.Spec.InfrastructureRef == nil {
+		return nil
+	}
+
+	// Call generic external reconciler.
+	infraConfig, err := r.reconcileExternal(ctx, cluster, cluster.Spec.InfrastructureRef)
+	if infraConfig == nil && err == nil {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	if cluster.Status.InfrastructureReady || !infraConfig.GetDeletionTimestamp().IsZero() {
```

Comment:
Should we split this and move the infra ready check up above the call to r.reconcileExternal?

Reply:
This looks fine to me, if we return early above then we don't reconcile the errors. I think the Machine `reconcileBootstrap` method needs to be fixed as well

Reply:
What flow are you thinking about? Cluster infraReady=true, but something possibly has changed to cause reconcileExternal to fail?